src/common/conn/server.rs

```diff
@@ -222,54 +261,110 @@ where
         exec: Executor<'a>,
+        app_connector: C,
+        app_config: AppClientConfig,
     ) -> Self {
         Self {
             http_listener,
             webrtc_config,
             exec,
+            app_connector,
+            app_config,
+            app_client: None,
+            webtrc_conn: None,
         }
     }
     pub fn serve_forever(&mut self, robot: Arc<Mutex<LocalRobot>>) {
         let cloned_exec = self.exec.clone();
         block_on(cloned_exec.run(Box::pin(self.serve(robot))));
     }
     async fn serve(&mut self, robot: Arc<Mutex<LocalRobot>>) {
+        let conn = self.app_connector.connect().unwrap();
+        let cloned_exec = self.exec.clone();
+        let grpc_client =
+            Box::new(GrpcClient::new(conn, cloned_exec, "https://app.viam.com:443").unwrap());
+        let app_client = AppClientBuilder::new(grpc_client, self.app_config.clone())
+            .build()
+            .unwrap();
+        let _ = self.app_client.insert(app_client);
+        let cloned_robot = robot.clone();
         loop {
             let _ = smol::Timer::after(std::time::Duration::from_millis(100)).await;
+            let sig = if let Some(webrtc_config) = self.webrtc_config.as_ref() {
+                let ip = self.app_config.get_ip();
+                let signaling = self.app_client.as_mut().unwrap().connect_signaling();
+                futures::future::Either::Left(WebRTCSignalingAnswerer {
+                    webrtc_config: Some(webrtc_config),
+                    future: signaling,
+                    ip,
+                })
+            } else {
+                futures::future::Either::Right(WebRTCSignalingAnswerer::<
+                    '_,
+                    '_,
+                    CC,
+                    D,
+                    futures_lite::future::Pending<Result<AppSignaling, AppClientError>>,
+                >::default())
+            };
+
             let listener = self.http_listener.next_conn();
-            let webrtc = self.webrtc_config.connect_webrtc().await;
+
             log::info!("waiting for connection");
+
             let connection = futures_lite::future::or(
                 async move {
                     let p = listener.await;
                     p.map(IncomingConnection::Http2Connection)
                         .map_err(|e| ServerError::Other(e.into()))
                 },
-                async move {
-                    let wrt = webrtc.connect().await;
-                    wrt.map(IncomingConnection::WebRtcConnection)
-                        .map_err(|e| ServerError::Other(e.into()))
+                async {
+                    let mut api = sig.await.map_err(|e| ServerError::Other(Box::new(e)))?;
+                    if let Some(task) = self.webtrc_conn.as_ref() {
+                        if !task.is_finished() {
+                            log::info!(
+                                "a webrtc connection is active ignoring further signaling requests"
```

Comment:
So, is there roughly where, in a future where we have priorities attached to connections, we can make a decision about whether to accept multiple, or drop lower priority if constrained?

Reply:
indeed